File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let maxProfit = 0;
    let left = 0;
    let right = 1;
    
    while (right < prices.length) {
        if (prices[left] < prices[right]) {
            const currentProfit = prices[right] - prices[left];
            if (currentProfit > maxProfit) {
                maxProfit = currentProfit;
            }
        } else {
            left = right;
        }
        right++;
    }
    return maxProfit;
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l, r = 0, 1
        maxProfit = 0
        
        while r < len(prices):
            currProfit = prices[r] - prices[l]
            if prices[l] > prices[r]:
                l = r
            else:
                maxProfit = max(currProfit, maxProfit)
            r += 1
            
        return maxProfit